### ISSUE TITLE
Incomplete distribution of dinucleotide shuffles

### ISSUE THREAD

**Issue by d-laub** [jmschrei/tangermeme#27]
First, thanks for creating tangermeme, it's very useful!

I have historically used the [uShuffle](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-192) algorithm for dinucleotide shuffling, which I have reimplemented in Rust/SeqPro for parallelization, so I wanted to see how they compared to the implementation in tangermeme. On a short 13 bp sequence for testing, I noticed a discrepancy where tangermeme returned fewer unique shuffles.

```python
# imports
from tangermeme import ersatz
import tangermeme as tg
from ushuffle import shuffle
import torch
import numpy as np
import seqpro as sp

# setup
seq = b'ACCCACGATGATG'
n_shuffles = int(1e5)

# using uShuffle reference implementation
>>> len({shuffle(seq, 2) for _ in range(n_shuffles)})
27

# tangermeme
>>> ohe = tg.utils.one_hot_encode(seq.decode())[None]
>>> tg_shuffles = ersatz.dinucleotide_shuffle(ohe, n=n_shuffles)[0]
>>> len({tg.utils.characters(s) for s in tg_shuffles})
9

# seqpro
>>> sp_shuffles = sp.k_shuffle(np.tile(np.asarray(seq), n_shuffles), 2, length_axis=-1)
>>> len(np.unique(sp_shuffles.view('S13')))
27
```

Is this a bug in the tangermeme implementation?

**Comment by jmschrei** [comment]
Yes -- this has actually already been fixed on my local branch but I have had a few deadlines in a row and haven't yet pushed the changes. If you pass in `end=None` that should solve the issue until I push it. 

There are actually two issues. The first is that the `dinucleotide_shuffle` function accepts a `start` and `end` argument and has `end` be -1 by default.. which means that the last nucleotide isn't included in the shuffle. Whoops.

The second issue is that the dinucleotide shuffle code I've used (not sure about you) uses a Eulerian path which requires that the first and last nucleotides are the same so that greedy decoding works. The problem is that many of these implementations keep the final transition for each nucleotide the same (i.e., if the last A transitions to a T in the original sequence, the last A will always transition to a T in the shuffles sequences... etc for all four nucleotides). This severely restricts the space. In one recent test case we found that, for a 20bp sequence, of ~220 valid dinucleotide shuffles found through brute force, these Eulerian path algorithms only found 8. If you're doing something similar you might want to look out for that, as well. This component is not yet included in the branch as I'm not sure what the best approach for fixing it is.

**Comment by d-laub** [comment]
Hi Jacob,

Thanks for the quick reply. I tried changing end=None in the tangermeme implementation and got the same result of 9 unique shuffles. Also, if I understand correctly, the unique sequences returned from uShuffle and SeqPro do not keep the final transition. The original transition in the above test case is T->G and 9/27 of them end in C->G. Perhaps this was addressed in the uShuffle implementation?

**Comment by d-laub** [comment]
Following up on the behavior you described, my guess is that it's due to this [line](https://github.com/jmschrei/tangermeme/blob/main/tangermeme/ersatz.py#L461) in ersatz._fast_shuffle. The algorithm implemented there seems to correspond to [Altschul & Erickson 1985](https://academic.oup.com/mbe/article/2/6/526/981780). Jiang et al. state in the uShuffle paper that "... a crucial step of their algorithm for generating random arborescences depends on a trial-and-error procedure, which is a potential bottleneck in performance." I think this refers to step 4 in Altschul & Erickson's algorithm:
1. Construct the doublet graph $G$ and edge ordering $E$ corresponding to $S$.
2. For each vertex $s$ in $G$ except $s_f$, randomly select one edge from the $s$ edge
list of $E(S)$ to be the last edge of the $s$ list in a new edge ordering.
3. From this set of last edges, construct the last-edge graph $Z$ and determine
whether or not all of its vertices are connected to $s_f$.
4. If any vertex is not connected in $Z$ to $s_f$, the new edge ordering will not be
Eulerian, so return to (2). If all vertices are connected in $Z$ to $s_f$, the new edge ordering
will be Eulerian, so continue to (5).
5. For each vertex $s$ in $G$, randomly permute the remaining edges of the $s$ edge
list of $E(S)$ to generate the $s$ edge list of the new edge ordering $E(S)$.
6. Construct sequence $S’$, a random DP permutation of $S$, from $E(S’)$ as follows.
Start at the $s_1$ edge list. At each $s_i$ edge list, add $s_i$ to $S’$, delete the first edge $S_i S_j$ of the
edge list, and move to the $s_i$ edge list. Continue this process until all edge lists are
exhausted.

In the aforementioned [line](https://github.com/jmschrei/tangermeme/blob/main/tangermeme/ersatz.py#L461) of ersatz._fast_shuffle, I think the implementation avoids this trial-and-error step by keeping the last edge (the last transition) the same as the input sequence, thereby guaranteeing the test condition in step 4. If I haven't misunderstood things somewhere, I think this is an artifact of this specific shortcut for Altschul and Erickson's algorithm, rather than a general issue with Eulerian-path algorithms for k-let frequency preserving shuffling.

**Comment by caenrigen** [comment]
Hi Jacob,

It does seem to solve the issue reported in this thread 🙂 here is a quick script results:

```python
from tangermeme import ersatz
from dinuc_shuf import shuffle as dinuc_shuf_seqs
import time
import tangermeme as tg
import torch

seq = b"ACCCACGATGATA"
n_shuffles = int(1e5)
ohe = tg.utils.one_hot_encode(seq.decode())[None]
assert ohe.shape == (1, 4, len(seq))

t0 = time.time()
tg_shuffles = ersatz.dinucleotide_shuffle(ohe, n=n_shuffles, end=-1)[0]
t1 = time.time()
delta = t1 - t0

assert tg_shuffles.shape == (n_shuffles, 4, len(seq))
unique_shufs = len({tg.utils.characters(s) for s in tg_shuffles})

print(f"tangermeme: {unique_shufs} unique shuffles, {delta:.3g} seconds")

ohe_np = ohe.squeeze().transpose(1, 0).numpy()
tiled = np.tile(ohe_np, (n_shuffles, 1, 1))
assert tiled.shape == (n_shuffles, len(seq), 4)

t0 = time.time()
shufs = dinuc_shuf_seqs(tiled, verify=True)
t1 = time.time()
delta = t1 - t0

shufs = torch.from_numpy(shufs).transpose(1, 2)
assert shufs.shape == (n_shuffles, 4, len(seq))
unique_shufs = len({tg.utils.characters(s) for s in shufs})

print(f"dinuc_shuf_seqs: {unique_shufs} unique shuffles, {delta:.3g} seconds")
```

Which outputs:

```bash
tangermeme: 9 unique shuffles, 0.0878 seconds
dinuc_shuf_seqs: 72 unique shuffles, 0.604 seconds
```

I don't know however how fast this implementation really is. Here it seems slower than `tangermeme` but it is not a fair comparison since the outputs are not complete.

I wanted to do a quick benchmark against `ushuffle` and `seqpro` but I am not able to install neither of these on my machine (macOS arm64 python 13.3.2), both run into compilation errors.

**Comment by jmschrei** [comment]
Thanks for following up, this is super useful. Would you mind checking how long this takes for sequences of varing length, ranging from the 10 to 1,000,000 bp? In principle I'd be happy to at least include a "complete" option which may be slower but covers all possible dinucleotide shuffles, but it'd be good to know how much slower it ends up being -- particulary for long sequences.

**Comment by caenrigen** [comment]
Actually looks pretty good with a bit of stats:

```python

from tangermeme.utils import random_one_hot
import pandas as pd
from tqdm.auto import tqdm
from collections import defaultdict

stats_tg = defaultdict(float)
stats_dinuc_shuf = defaultdict(float)

n_shuffles = 30
repeats = 30
for power in tqdm(range(1, 6 + 1)):
    seq_len = 10**power
    X = random_one_hot((1, 4, seq_len), random_state=413).type(torch.float32)
    X_tiled = np.tile(X.squeeze().transpose(1, 0).numpy(), (n_shuffles, 1, 1))

    for _ in tqdm(range(repeats), leave=False):
        t0 = time.time()
        tg_shuffles = ersatz.dinucleotide_shuffle(X, n=n_shuffles, random_state=413)[0]
        t1 = time.time()
        stats_tg[power] += t1 - t0

        t0 = time.time()
        shufs = dinuc_shuf_seqs(X_tiled, verify=False)
        t1 = time.time()
        stats_dinuc_shuf[power] += t1 - t0

    stats_tg[power] /= repeats
    stats_dinuc_shuf[power] /= repeats

df = pd.DataFrame({"tangermeme": stats_tg, "dinuc_shuf": stats_dinuc_shuf})
df.set_index(10**df.index, inplace=True)
print(df)

df.plot(
    marker="o",
    logx=True,
    logy=True,
    xlabel="Sequence length",
    ylabel=f"Time {n_shuffles} shuffles(s)",
)
```

```
         tangermeme  dinuc_shuf
10         0.000350    0.000303
100        0.000539    0.000345
1000       0.001480    0.001307
10000      0.013387    0.011981
100000     0.118063    0.126649
1000000    1.377906    1.285338
```

<img width="578" height="437" alt="Image" src="https://github.com/user-attachments/assets/606be7a4-514f-4f11-9458-a096d2f60f62" />

**Comment by d-laub** [comment]
Thanks for sharing dinuc-shuf @caenrigen, I wasn't aware of that implementation and it looks really good. I added SeqPro to your micro-benchmark (and uploaded wheels for windows and mac).

```python
import torch
import time
import numpy as np
from tangermeme.utils import random_one_hot
import pandas as pd
from tqdm.auto import tqdm
from collections import defaultdict
from dinuc_shuf import shuffle as dinuc_shuf_seqs
from tangermeme import ersatz
import seqpro as sp

def bench_kshuffle(n_shuffles: int, repeats: int, max_power: int = 6) -> pd.DataFrame:
    stats_tg = defaultdict(float)
    stats_dinuc_shuf = defaultdict(float)
    stats_sp = defaultdict(float)

    for power in tqdm(range(1, max_power + 1)):
        seq_len = 10**power
        X = random_one_hot((1, 4, seq_len), random_state=413).type(torch.float32)
        X_tiled = np.tile(X.squeeze().transpose(1, 0).numpy(), (n_shuffles, 1, 1))
        X_tiled_bytes = sp.DNA.decode_ohe(X_tiled.astype(np.uint8), ohe_axis=-1)

        for _ in tqdm(range(repeats), leave=False):
            t0 = time.time()
            tg_shuffles = ersatz.dinucleotide_shuffle(X, n=n_shuffles, random_state=413)[0]
            t1 = time.time()
            stats_tg[seq_len] += t1 - t0

            t0 = time.time()
            shufs = dinuc_shuf_seqs(X_tiled, verify=False)
            t1 = time.time()
            stats_dinuc_shuf[seq_len] += t1 - t0

            t0 = time.time()
            sp_shufs = sp.k_shuffle(X_tiled_bytes, k=2, length_axis=-1)
            t1 = time.time()
            stats_sp[seq_len] += t1 - t0

        stats_tg[seq_len] /= repeats
        stats_dinuc_shuf[seq_len] /= repeats
        stats_sp[seq_len] /= repeats

    df = pd.DataFrame(
        {"tangermeme": stats_tg, "dinuc_shuf": stats_dinuc_shuf, "seqpro": stats_sp}
    )
    return df

n_shuffles = 30
repeats = 30
df = bench_kshuffle(n_shuffles, repeats)
print(df)

df.plot(
    marker="o",
    logx=True,
    logy=True,
    xlabel="Sequence length",
    ylabel=f"Time {n_shuffles} shuffles(s)",
)
```

<img width="573" height="436" alt="Image" src="https://github.com/user-attachments/assets/29ca794b-3cee-4c7c-9b30-c59775bba939" />

```python
n_shuffles = 120
repeats = 10
df = bench_kshuffle(n_shuffles, repeats)
print(df)

df.plot(
    marker="o",
    logx=True,
    logy=True,
    xlabel="Sequence length",
    ylabel=f"Time {n_shuffles} shuffles(s)",
)
```

<img width="573" height="441" alt="Image" src="https://github.com/user-attachments/assets/c3f38b16-0226-4d29-84af-881977720744" />

It looks like dinuc-shuf might be an ideal default implementation and, as a stand-alone package, it should be quite painless to add as a dependency. It also looks like dinuc-shuf is not using multithreading in batched_shuffle, so this could potentially (and very easily) be accelerated proportional to the number of available cores.

**Comment by d-laub** [comment]
Optimized seqpro a bit (v0.11.1), here's an updated bench. [Gist here](https://gist.github.com/d-laub/e33b023a71f7759da915e8810e8b238d).

<img width="808" height="489" alt="Image" src="https://github.com/user-attachments/assets/783ad540-a889-4aae-b0e6-108a0c073b81" />
